src/shared/Atlassian.Bitbucket.Tests/BitbucketOauth2ClientTest.cs

```diff
@@ -0,0 +1,142 @@
+
+namespace Atlassian.Bitbucket.Tests
+{
+    public class BitbucketOAuth2ClientTest
+    {
+        private Mock<HttpClient> httpClient = new Mock<HttpClient>(MockBehavior.Strict);
+        private Mock<ISettings> settings = new Mock<ISettings>(MockBehavior.Loose);
+        private Mock<IOAuth2WebBrowser> browser = new Mock<IOAuth2WebBrowser>(MockBehavior.Strict);
+        private Mock<IOAuth2CodeGenerator> codeGenerator = new Mock<IOAuth2CodeGenerator>(MockBehavior.Strict);
+        private IEnumerable<string> scopes = new List<string>();
+        private CancellationToken ct = new CancellationToken();
+        private Uri rootCallbackUri = new Uri("http://localhost:34106/");
+        private string nonce = "12345";
+        private string pkceCodeVerifier = "abcde";
+        private string pkceCodeChallenge = "xyz987";
+        private string authorization_code = "authorization_token";
+
+        [Fact]
+        public async Task BitbucketOAuth2Client_GetAuthorizationCodeAsync_ReturnsCode()
+        {
+            MockClientIdOverride(false, "never used");
+
+            Uri finalCallbackUri = MockFinalCallbackUri();
+
+            MockGetAuthenticationCodeAsync(finalCallbackUri, null);
+
+            MockCodeGenerator();
+
+            BitbucketOAuth2Client client = GetBitbucketOAuth2Client();
+
+            var result = await client.GetAuthorizationCodeAsync(scopes, browser.Object, ct);
+
+            VerifyAuthorizationCodeResult(result);
+        }
+
+        [Theory]
+        [InlineData(null)]
+        [InlineData("i234")]
+        public async Task BitbucketOAuth2Client_GetAuthorizationCodeAsync_RespectsClientIdOverride_ReturnsCode(string clientId)
+        {
+            MockClientIdOverride(clientId != null, clientId);
+
+            Uri finalCallbackUri = MockFinalCallbackUri();
+
+            MockGetAuthenticationCodeAsync(finalCallbackUri, clientId);
+
+            MockCodeGenerator();
+
+            BitbucketOAuth2Client client = GetBitbucketOAuth2Client();
+
+            var result = await client.GetAuthorizationCodeAsync(scopes, browser.Object, ct);
+
+            VerifyAuthorizationCodeResult(result);
+        }
+
+        [Fact]
+        public async Task BitbucketOAuth2Client_GetDeviceCodeAsync()
+        {
+            var client = new BitbucketOAuth2Client(httpClient.Object, settings.Object);
+            await Assert.ThrowsAsync<InvalidOperationException>(async () => await client.GetDeviceCodeAsync(scopes, ct));
+        }
+
+        private void VerifyOAuth2TokenResult(OAuth2TokenResult result)
+        {
+            Assert.NotNull(result);
+            IEnumerable<char> access_token = null;
+            Assert.Equal(access_token, result.AccessToken);
+            IEnumerable<char> refresh_token = null;
+            Assert.Equal(refresh_token, result.RefreshToken);
+            IEnumerable<char> tokenType = null;
+            Assert.Equal(tokenType, result.TokenType);
+            Assert.Equal(null, result.Scopes);
```

Comment:
Xunit prefers `Assert.Null(X)` over `Assert.Equal(null, X)`:

```suggestion
            Assert.Null(result.Scopes);
```